processor/src/debug.rs

```diff
@@ -313,3 +314,113 @@ impl AsRef<AssemblyOp> for AsmOpInfo {
         &self.asmop
     }
 }
+
+// BUS DEBUGGING
+// =================================================================
+
+/// A message that can be sent on a bus.
+pub(crate) trait BusMessage<E: FieldElement<BaseField = Felt>>: fmt::Display {
+    /// The concrete value that this message evaluates to.
+    fn value(&self, alphas: &[E]) -> E;
+
+    /// The source of this message (e.g. "mload" or "memory chiplet").
+    fn source(&self) -> &str;
+}
+
+/// Note: we use `Vec` internally instead of a `BTreeMap` as a workaround for field elements not
+/// implementing `Ord`. Since this is only used in debug/test code, this is acceptable.
+pub(crate) struct BusDebugger<E: FieldElement<BaseField = Felt>> {
```

Comment:
We could convert field element to string and use the string as a key in a `BTreeMap`.

Reply:
While trying it out I noticed that the type we'd really need is

```rs
pub outstanding_{requests, responses}: BTreeMap<String, Vec<Box<dyn BusMessage<E>>>>,
```

since there can be collisions in messages. For example, the bitwise chiplet messages can collide when you request the same `and` or `or` (since the `clk` is not included in the message). This was exercised by the `b_chip_trace_bitwise()` test.

This results in relatively complex insertion/removal logic. Since this is test code, I prefer readability over performance, so I think we should leave it as is. For now I adjusted the comment to reflect this.

Reply:
Ah - makes sense. We'll probably have the same situation when we deal with LogUp-based busses.